data_steward/analytics/cdr_ops/incremental_load_qc.py

```diff
@@ -0,0 +1,348 @@
+    GROUP BY table_name, observation_id
+    HAVING COUNT(*) > 1
+''')
+
+queries = [
+    query.render(project=project_id, new_dataset=new_dataset, table=table)
+    for table in obs_tables
+]
+
+union_all_query = '\nUNION ALL\n'.join(queries)
+df = execute(client, union_all_query)
+
+success_null_check = f"No OBSERVATION_ID duplicates are found in {new_dataset}.<br><br>"
+failure_null_check = (
+    "There are total <b>{count}</b> OBSERVATION_IDs that might have"
+    "duplicates. <br>Look at these tables and investigate why there are duplicates: <b>{tables}</b><br><br>"
+)
+
+render_message(df,
+               success_null_check,
+               failure_null_check,
+               failure_msg_args={
+                   'count': len(df),
+                   'tables': ', '.join(df.table_name)
+               })
+# -
+
+# ## QC 3. Confirm mapping/ext tables are consistent
+# This hotfix runs several delete and insert statements to OBSERVATION, SURVEY_CONDUCT, PERSON
+# and their mapping/ext tables. <br>We must confirm that mapping/ext tables are consistent with
+# their correspondants after the hotfix.
+
+# +
+map_ext_tuples = [
+    ('observation', 'observation_ext'),
+    ('survey_conduct', 'survey_conduct_ext'),
+    ('person', 'person_ext'),
+] if is_deidentified == 'True' else [
+    ('observation', '_mapping_observation'),
+    ('observation', 'observation_ext'),
+    ('survey_conduct', '_mapping_survey_conduct'),
+    ('survey_conduct', 'survey_conduct_ext'),
+]
+
+query = JINJA_ENV.from_string('''
+    SELECT
+        \'{{mapping_table}}\' as table_name,
+        COUNT(*) AS unmatched_ids
+    FROM `{{project}}.{{new_dataset}}.{{table_name}}` d
+    FULL OUTER JOIN `{{project}}.{{new_dataset}}.{{mapping_table}}` mp
+    USING ({{table_name}}_id)
+    WHERE d.{{table_name}}_id IS NULL OR mp.{{table_name}}_id IS NULL
+    HAVING COUNT(*) > 0
+''')
+
+queries = []
+for (table, mapping_table) in map_ext_tuples:
+    queries.append(
+        query.render(project=project_id,
+                     new_dataset=new_dataset,
+                     table_name=table,
+                     mapping_table=mapping_table))
+union_all_query = '\nUNION ALL\n'.join(queries)
+
+df = execute(client, union_all_query)
+
+success_null_check = (
+    "All OBSERVATION, SURVEY_CONDUCT, and PERSON records have "
+    "valid corresponding records in their mapping/ext tables. <br>"
+    "And there are no invalid records in the mapping/ext tables.")
+failure_null_check = (
+    "These mapping/ext tables are inconsistent with their "
+    "correspondants. <br>Look at these tables and investigate why "
+    "they are inconsistent: <b>{tables}</b><br><br>")
+
+render_message(df,
+               success_null_check,
+               failure_null_check,
+               failure_msg_args={'tables': ', '.join(df.table_name)})
+
+# -
+
+# ## QC 4. (Only for `is_deidentified == True`) Confirm `person_ext`'s state related columns come from `source_dataset`
```

Comment:
@hiro-mishima Molly and I like this check but noticed it is checking the entire column values from the source dataset to the entire column values from the new dataset.  This is a good check.  We also suggest running this check limited to the scope of the largest possible change (i.e. for just the participants in the incremental dataset).  Then we could know if an issue is found between the two if it is from the original data or from the increment or both.

Reply:
Thanks! I added the additional check with this commit: https://github.com/all-of-us/curation/pull/1489/commits/5990a2ee67197f319de27e9d16db03baa0b9e77b